File: ShortPath/ShortPath.py
import numpy as np
import sys

#Cria os vertices no grafo, 
vertices = {'arad': 0, 'zerind': 1, 'oradea': 2, 'sibiu': 3, 'timisoara': 4,
            'lugoj': 5, 'mehadia': 6, 'dobreta': 7, 'craiova': 8, 'rimnicu': 9,
            'fagaras': 10, 'pitesti': 11, 'bucharest': 12, 'giurgiu': 13}

# Cria as arestas no grafo
cidades = {0: 'arad', 1: 'zerind', 2: 'oradea', 3: 'sibiu', 4: 'timisoara',
           5: 'lugoj', 6: 'mehadia', 7: 'dobreta', 8: 'craiova', 9: 'rimnicu',
           10: 'fagaras', 11: 'pitesti', 12: 'bucharest', 13: 'giurgiu'}

arestas = np.zeros([len(cidades), len(cidades)], dtype=int)

class Dijkstra:
    def __init__(self, vertices, arestas, inicio): # Função Construtor
        self.tamanho = len(vertices)
        self.vertices = vertices
        self.grafo = arestas
        self.inicio = inicio

    def MostraSolucao(self, distancias): #Exibi na tela a solução do menor caminho
        print('Menores distâncias de {} ate todos os outros'.format(self.vertices[self.inicio]))
        for vertice in range(self.tamanho):
            print(self.vertices[vertice], distancias[vertice])  

    def distanciaMinima(self, distancia, visitados): #Mostra a distancia minima
        minimo = sys.maxsize
        for vertice in range(self.tamanho):
            if distancia[vertice] < minimo and visitados[vertice] == False:
                minimo = distancia[vertice]
                indiceMinimo = vertice
        return indiceMinimo

    def dijkstra(self): #Verifica qual é menor caminho
        distancia = [sys.maxsize] * self.tamanho
        distancia[self.inicio] = 0
        visitados = [False] * self.tamanho

        for _ in range(self.tamanho):
            indiceMinimo = self.distanciaMinima(distancia, visitados)
            visitados[indiceMinimo] = True
            for vertice in range(self.tamanho):
                if self.grafo[indiceMinimo][vertice] > 0 and visitados[vertice] == False and distancia[vertice] > distancia[indiceMinimo] + self.grafo[indiceMinimo][vertice]:
                    distancia[vertice] = distancia[indiceMinimo] + self.grafo[indiceMinimo][vertice]

        self.MostraSolucao(distancia)

                    
#Adiciona os valores na matriz
arestas[vertices['arad'], [vertices['zerind']]] = 75
arestas[vertices['arad'], [vertices['sibiu']]] = 140
arestas[vertices['arad'], [vertices['timisoara']]] = 118

arestas[vertices['zerind'],[vertices['arad']]] = 75
arestas[vertices['zerind'],[vertices['oradea']]] = 71

arestas[vertices['oradea'],[vertices['zerind']]] = 71
arestas[vertices['oradea'],[vertices['sibiu']]] = 151

arestas[vertices['sibiu'],[vertices['oradea']]] = 151
arestas[vertices['sibiu'],[vertices['arad']]] = 140
arestas[vertices['sibiu'],[vertices['fagaras']]] = 99
arestas[vertices['sibiu'],[vertices['rimnicu']]] = 80

arestas[vertices['timisoara'],[vertices['arad']]] = 118
arestas[vertices['timisoara'],[vertices['lugoj']]] = 111

arestas[vertices['lugoj'],[vertices['timisoara']]] = 111
arestas[vertices['lugoj'],[vertices['mehadia']]] = 70

arestas[vertices['mehadia'],[vertices['lugoj']]] = 70
arestas[vertices['mehadia'],[vertices['dobreta']]] = 75

arestas[vertices['dobreta'],[vertices['mehadia']]] = 75
arestas[vertices['dobreta'],[vertices['craiova']]] = 120

arestas[vertices['craiova'],[vertices['dobreta']]] = 120
arestas[vertices['craiova'],[vertices['pitesti']]] = 138
arestas[vertices['craiova'],[vertices['rimnicu']]] = 146

arestas[vertices['rimnicu'],[vertices['craiova']]] = 146
arestas[vertices['rimnicu'],[vertices['sibiu']]] = 80
arestas[vertices['rimnicu'],[vertices['pitesti']]] = 97

arestas[vertices['fagaras'],[vertices['sibiu']]] = 99
arestas[vertices['fagaras'],[vertices['bucharest']]] = 211

arestas[vertices['pitesti'],[vertices['rimnicu']]] = 97
arestas[vertices['pitesti'],[vertices['craiova']]] = 138
arestas[vertices['pitesti'],[vertices['bucharest']]] = 101

arestas[vertices['bucharest'],[vertices['fagaras']]] = 211
arestas[vertices['bucharest'],[vertices['pitesti']]] = 101
arestas[vertices['bucharest'],[vertices['giurgiu']]] = 90

dijkstra = Dijkstra(cidades, arestas, vertices['arad'])
dijkstra.dijkstra()

File: ShortPath/ShortPath.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;
constexpr int MAXNOS = 15;
constexpr int INFINITO = 1000000;

class Mapa{
  public:
    Mapa();
    void addCidade(string nomeCidade);
    void une(string cidadeA, string cidadeB, float distancia);
    void menorCaminho(string cidadeA, string cidadeB);

  private:
    int N;
    string cidades[MAXNOS];
    bool adj[MAXNOS][MAXNOS];
    float peso[MAXNOS][MAXNOS];
    int indiceCidade(string nomeCidade);
    float distanciaCidade(int corrente, int i);
    bool validos(int a, int b);
};

Mapa::Mapa(){
  N = 0;

  for(int i = 0; i < MAXNOS; ++i){
    for(int j = 0; j < MAXNOS; ++j){
      adj[i][j] = false;
      peso[i][j] = INFINITO;
    }
  }
}

int Mapa::indiceCidade(string nomeCidade){
  for(int i = 0; i < N; ++i){
    if(cidades[i] == nomeCidade)
      return i;
  }

  return -1;
}

bool Mapa::validos(int a, int b){
  return (a > -1 && a < MAXNOS && b > -1 && b < MAXNOS);
}

void Mapa::addCidade(string nomeCidade){
  if(N < MAXNOS){
    cidades[N] = nomeCidade;
    ++N;
  }else
    cerr << "Numeros maximo atingindo!" << endl;
}

void Mapa::une(string cidadeA, string cidadeB, float distancia){
  if(N > 0){
    int a = indiceCidade(cidadeA);
    int b = indiceCidade(cidadeB);

    if(validos(a, b)){
      adj[a][b] = adj[b][a] = true;
      peso[a][b] = peso[b][a] = distancia;
    }
  }
}

float Mapa::distanciaCidade(int atual, int i){
  if(validos(atual, i)){
    if(adj[i][atual])
      return peso[i][atual];
  }else
    return INFINITO;
}

void Mapa::menorCaminho(string cidadeA, string cidadeB){
  int s = indiceCidade(cidadeA);
  int t = indiceCidade(cidadeB);
  int precede[MAXNOS] = {s};
  string menorCaminho[MAXNOS];
  float distancia[MAXNOS];
  bool calculado[MAXNOS];

  for(int i = 0; i < N; ++i){
    distancia[i] = INFINITO;
    calculado[i] = false;
  }

  distancia[s] = 0;
  calculado[s] = true;

  int atual = s;
  while(atual != t){
    float menorDist = INFINITO;
    int k = 0;
    float dc = distancia[atual];
    for(int i = 0; i < N; ++i){
      if(!calculado[i]){
        float novaDist = dc + distanciaCidade(atual, i);
        if(novaDist < distancia[i]){
          distancia[i] = novaDist;
          precede[i] = atual;
        }

        if(distancia[i] < menorDist){
          menorDist = distancia[i];
          k = i;
        }
      }
    }
    atual = k;
    calculado[atual] = true;
  }

  string caminho = cidades[t];
  int caminhoInt = t;
  int cont = 0;

  while(caminho != cidadeA){
    menorCaminho[cont] = caminho;
    caminho = cidades[precede[caminhoInt]];
    caminhoInt = indiceCidade(caminho);
    ++cont;
  }

  for(int i = 0, j = cont - 1; i < cont / 2; i++, j++){
    string temp = menorCaminho[i];
    menorCaminho[i] = menorCaminho[j];
    menorCaminho[j] = temp;
  }

  cout << "Menor Distancia: " << distancia[t] << "Km" << endl;
  cout << "Menor Caminho: ";
  cout << cidadeA << " -> ";
  for(int i = 0; i < cont; i++){
    cout << menorCaminho[i];
    if(i != cont - 1)
      cout << " -> ";
  }
}

int main(){
  Mapa m;

  m.addCidade("Arad");
  m.addCidade("Zerind");
  m.addCidade("Oradea");
  m.addCidade("Sibiu");
  m.addCidade("Timisoara");
  m.addCidade("Lugoj");
  m.addCidade("Mehadia");
  m.addCidade("Dobreta");
  m.addCidade("Craiova");
  m.addCidade("Rimnicu");
  m.addCidade("Fagaras");
  m.addCidade("Pitesti");
  m.addCidade("Bucharest");
  m.addCidade("Giurgiu");

  m.une("Arad", "Zerind", 271.0);
  m.une("Arad", "Oradea", 310.0);
  m.une("Oradea", "Zerind", 174.0);
  m.une("Oradea", "Sibiu", 208.0);
  m.une("Sibiu", "Timisoara", 342.0);
  m.une("Sibiu", "Lugoj", 114.0);
  m.une("Lugoj", "Mehadia", 115.0);
  m.une("Lugoj", "Sibiu", 132.0);
  m.une("Mehadia", "Dobreta", 221.2);
  m.une("Mehadia", "Craiova", 185.9);
  m.une("Mehadia", "Rimnicu", 155.0);
  m.une("Craviova", "Fagaras", 143.0);
  m.une("Craviova", "Mehadia", 165.0);
  m.une("Fagaras", "Rimnicu", 136.0);
  m.une("Fagaras", "Pitesti", 209.0);
  m.une("Fagaras", "Bucharest", 271.0);
  m.une("Fagaras", "Craviova", 349.0);
  m.une("Pitesti", "Bucharest", 223.0);
  m.une("Buchasrest", "Pitesti", 231.0);
  m.une("Buchasrest", "Giurgiu", 102.0);
  m.une("Rimnicu", "Fagasras", 52.3);

  string cidadeA, cidadeB;

  cout << "Cidade A: ";
  getline(cin, cidadeA);
  cout << "Cidade B: ";
  getline(cin, cidadeB);
  cout << endl;
  m.menorCaminho(cidadeA, cidadeB);
  cout << endl;

  return 0;
}